Instrument identity, registration conflicts, multi-callback registration, unregister support (#2317)
open-telemetry/opentelemetry-specification: specification/metrics/api.md

* [since 1.4.0] `schema_url` (optional): Specifies the Schema URL that should be
  recorded in the emitted telemetry.

Meters are identified by all of these fields.  When more than one
Meter of the same `name`, `version`, and `schema_url` is created, it

> is unspecified whether or under which conditions the same or different

Reviewer: I think erlang is the poster child for that iirc. In order to have the same instance returned there needs to be some state maintained external to the function which breaks the functional design principle.

`Meter` instances are returned.  Meter identity, and the terms
_identical_ and _distinct_ applied to Meters in this specification

> describes Meter instances that have identical and/or distinct values

Reviewer 2: I think this sentence needs to be redacted, if `and` and `or` are options, it may be read as `Meter instances that have identical and distinct values...` which sounds contradictory.

least one identifying field has a different value.

Implementations MUST NOT require users to repeatedly obtain a `Meter` with
the same identity to pick up configuration changes. This can be
achieved either by allowing to work with an outdated configuration or by

> ensuring that new configuration applies also to previously returned `Meter`s.

Reviewer: How do you define "working" if a meter can work with outdated configurations? If this is allowed, what is the non-working failure mode we are trying to guard against here?

Author: I'm a little uncertain about this (preexisting) statement. @reyang, do you think we could remove "work with an outdated configuration", or would you help us clarify its itent?

* An optional `description`

Instruments are associated with the Meter during creation. Instruments
are identified by all of these fields.

Language-level features such as the distinction between integer and
floating point numbers SHOULD be considered as identifying.

<a name="instrument-type-conflict-detection"></a>

When more than one Instrument of the same `name` is created for
identical Meters, denoted _duplicate instrument registration_, the
implementation MUST create a valid Instrument in every case.  Here,
"valid" means an instrument that is functional and can expected to
export data, despite potentially creating a [semantic error in the
data
model](datamodel.md#opentelemetry-protocol-data-model-producer-recommendations).

It is unspecified whether or under which conditions the same or
different Instrument instance will be returned as a result of
duplicate instrument registration.  The term _identical_ applied to
Instruments describes instances where all identifying fields are
equal.  The term _distinct_ applied to Instruments describes instances
where at least one field value is different.

The implementation MUST aggregate data from identical Instruments

> together in its export pipeline.

Author: Correct. I moved this line into sdk.md.

When more than one distinct Instrument is registered with the same
`name` for identical Meters, the implementation SHOULD emit a warning

> to the user informing them of duplicate registration conflict(s).

Author: The "warning" is supposed to include a Views recipe if possible, otherwise yes.

> will have the following information:

Reviewer 2: ```suggestion
will have the following fields:
```
```suggestion
will have the following information:
```

* The `name` of the Instrument
* The `kind` of the Instrument - whether it is a [Counter](#counter) or
  one of the other kinds, whether it is synchronous or asynchronous
* An optional `unit` of measure
* An optional `description`

Instruments are associated with the Meter during creation. Instruments
are identified by all of these fields.

<a name="instrument-type-conflict-detection"></a>

When more than one Instrument of the same `name` is created for
identical Meters, denoted _duplicate instrument registration_, the

> implementation MUST return a valid Instrument in every case.

Reviewer 3: We need to take `Hint API` in the consideration if we plan to add them in the near future.

If we allow _hints_ to be specified duration instrument creation, we'll have to handle multiple calls under a given Meter with the same Instrument name but different hints.

An alternative solution might be allowing _hints_ to be provided for a given instance of instrument - as long as there is way to get such instance. For example, an async instrument creation API might choose not to return a handle/reference.

Author: I've spelled out a lot more about instrument identity now, and in the data model we're left with some hard rules about what's identifying and what's not. Hints should lead to non-identifying traits, like histogram boundaries and scale for example.

> implementation MUST return a valid Instrument in every case.

Author: No. The word "valid" is still used. I've added a definition for what "Valid" means in this case.

It is unspecified whether or under which conditions the same or
different Instrument instance will be returned as a result of
duplicate instrument registration.  Instrument identity, and the terms
_identical_ and _distinct_ applied to Instruments in this
specification describes Instrument instances that have identical
and/or distinct values for all four of their `name`, `kind`, `unit`,

> and `description` attributes.  Language-level features such as the

Reviewer 4: In #2270 description was not part of instrument identity. Is this on purpose?

Author: Yes, it seems simpler than to create new exceptions anyway. See https://github.com/open-telemetry/opentelemetry-specification/issues/2226#issuecomment-1029527138

Reviewer 4: Sounds good! 👍

will have the following information:

* The `name` of the Instrument
* The `kind` of the Instrument - whether it is a [Counter](#counter) or 
  one of the other kinds, whether it is synchronous or asynchronous
* An optional `unit` of measure
* An optional `description`

Instruments are associated with the Meter during creation. Instruments
are identified by all of these fields.

<a name="instrument-type-conflict-detection"></a>

When more than one Instrument of the same `name` is created for
identical Meters, denoted _duplicate instrument registration_, the
implementation MUST return a valid Instrument in every case.

It is unspecified whether or under which conditions the same or
different Instrument instance will be returned as a result of
duplicate instrument registration.  Instrument identity, and the terms
_identical_ and _distinct_ applied to Instruments in this
specification describes Instrument instances that have identical
and/or distinct values for all four of their `name`, `kind`, `unit`,
and `description` attributes.  Language-level features such the

> distinction between integer and floating point numbers SHOULD be

Reviewer 5: Nit: This sentence feels off.

Language-level features such *as* the distinction...

Author: Fixed,

considered as identifying, as well.

The implementation MUST aggregate data from identical Instruments
together in its export pipeline.

> When through duplicate registration more than one distinct Instrument

Reviewer 5: Nit:  drop the "through duplicate registration", it's implied and makes this sentence hard to parse.

Author: Fixed.

* It MUST support at least 1023 characters. [OpenTelemetry
  API](../overview.md#api) authors MAY decide if they want to support more.

> Instruments are categorized based on whether they are synchronous or

Reviewer 2: ```suggestion
Instruments are categorized on whether they are synchronous or
```

with the [asynchronous
pattern](https://en.wikipedia.org/wiki/Asynchronous_method_invocation).

Considering duplicate registration of asynchronous instruments, the
implementation MUST support multiple callbacks registered per
identical instrument.

> The API MAY provide support for registration of a single callback

Author: This is a good example because `/proc/stat` contains many numbers that could be [useful to report](https://askubuntu.com/questions/120953/exact-field-meaning-of-proc-stat). The advantages of multiple instruments per callback are:

1. Read the /proc/stat file once, not once per metric
2. Counters and gauges are temporally aligned by nature (which is why the API spec says callbacks use one timestamp)

The second point means that you could read the CPU stats, procs_blocked, procs_running, softirq state, etc all at once and the receiver can interpret those observations as being correlated, by having the same timestamp. In the context of a [multivariate metrics protocol](https://github.com/open-telemetry/oteps/pull/171), I would define the events of one callback to appear as a single "row" in the columnar encoding => a receiver can study covariance of the variables across a fleet of /proc/stat readings.

Here's how I might write this example.

```python
cpuTime = meter.create_observable_counter("system.cpu.time")
procsRunning = meter.create_observable_gauge("system.procs_running")
procsBlocked = meter.create_observable_gauge("system.procs_blocked")

def proc_stat_observer() -> Iterable[Measurement]:
    with open("/proc/stat") as procstat:
        procstat.readline()  # skip the first line
        for line in procstat:
            if line.startswith("cpu"):
                cpu, *states = line.split()
                yield cpuTime.observe(int(states[0]) // 100, {"cpu": cpu, "state": "user"})
                yield cpuTime.observe(int(states[1]) // 100, {"cpu": cpu, "state": "nice"})
            if line.startswith("procs_running"):
                var, value = line.split()
                yield procsRunning.observe(value)
            if line.startswith("procs_blocked"):
                var, value = line.split()
                yield procsBlocked.observe(value)

callback = meter.register_callback(proc_stat_observer)

def stop():
    callback.unregister()
```

> The API MAY provide support for registration of a single callback

Reviewer 6: IIRC, the batch observer API was removed in the feature-freeze metrics API intentionally. Although, there are great use cases for it.

> having multiple asynchronous instruments, where idiomatic.  The

Reviewer 2: `having` is a bit unspecific here.

> implementation is REQUIRED to maintain a mapping from every regsitered

Reviewer 4: ```suggestion
implementation is REQUIRED to maintain a mapping from every registered
```

Author: Fixed, thanks.

> implementation is REQUIRED to maintain a mapping from every registered

Reviewer 3: This might need additional clarification; it sounds like we're suggesting that the callback function should be used as part of the identification.

Can a callback function be registered multiple times on a single instrument?

e.g.

```
var f = () => { /* something */ };
target.addEventListener("click", f);
target.addEventListener("click", f); // what would this line do?
target.removeEventListener("click", f);
```

or

```
var f = () => { /* something */ };
var atom1 = target.register("click", f);
var atom2 = target.register("click", f);
target.unregister(atom1); // or atom1.unregister();
target.unregister(atom2); // or atom2.unregister();
```

Reviewer: I can say that a typical JS event emitter would treat the first example as:

```javascript
var f = () => { /* something */ };
target.addEventListener("click", f);
target.addEventListener("click", f); // duplicate registration. emitting 'click' would call f twice
target.removeEventListener("click", f);
target.removeEventListener("click", f); // remove only one of the registrations. emitting 'click' would now call f once
```

Author: I still haven't added examples to this PR, sorry! Will catch up on Monday.

Author: See https://github.com/open-telemetry/opentelemetry-python/issues/2454

callback to the associated instrument(s).

> The implementation MUST provide a way to unregister callbacks

Reviewer 3: Do we want to mark this as `Experimental`, wait for prototypes across languages, then mark it as `Stable`?

Reviewer 2: Also, what should happen if an instrument's callback is called after it is unregistered?

Reviewer 3: @ocelotl I discussed this offline with @jmacd yesterday and he is working on clarifying these things.

associated with asynchronous instruments.

> The implementation SHOULD reject observations that are made outside of

Reviewer 6: This is an interesting point, +1 from me. But how do we define "outside of a callback"? like, say, for a single callback, when SDK initiates a new invocation of the callback, should an observation from a previous invocation of the callback be considered as "outside of a callback"?

Reviewer: For clarification this can happen if a callback takes a long time to complete and is still making observations when the next callback starts

Author: Good question, this needs to be better specified.

In a Golang setting, I would pass a `context.Context` with the timeout set during Collect.

1. Use `context.WithTimeout(...)`,
  - in a new goroutine for each instrumentation library in parallel,
  - for each registered callback in sequence,
  - execute the callback
2. In Collect(), wait for the timeout
  - allow canceled contexts to continue running asynchronously
  - do not accept observations on these contexts after the timeout
3. Collect synchronous instruments after asynchronous instruments
  - this keeps the synchronous data fresh in case the async callback(s) hit a deadline
  - question re: use of histogram instruments in asynchronous callbacks (e.g., https://github.com/open-telemetry/opentelemetry-specification/issues/607)
4. Process and pass data to the exporter

The SDK spec already has a lot of language about completing or aborting within timeouts, and states:

> `Collect` SHOULD provide a way to let the caller know whether it succeeded, failed or timed out.

The API spec has this for every asynchronous instrument:

> The callback function SHOULD NOT take indefinite amount of time. If multiple independent SDKs coexist in a running process, they MUST invoke the callback function(s) independently.

The API spec also states this, although it reads like an unstated SDK specification:

> The API MUST treat observations from a single callback as logically taking place at a single instant, such that when recorded, observations from a single callback MUST be reported with identical timestamps.

@legendecas @dyladan WDYT?

Reviewer 7: The java implementation takes a lock when calling async callbacks that doesn't release until the callback completes. This means there is no way to report observations for a previous invocation, since subsequent collects will be blocked until callbacks from previous collect all complete. It may be a good idea to limit the execution time of callbacks, but we haven't done that yet, and it does introduce more complexity.

Reviewer 3: > The java implementation takes a lock when calling async callbacks that doesn't release until the callback completes. This means there is no way to report observations for a previous invocation, since subsequent collects will be blocked until callbacks from previous collect all complete. It may be a good idea to limit the execution time of callbacks, but we haven't done that yet, and it does introduce more complexity.

@jack-berg is that a reentrant lock? Would it be possible for the callback to invoke itself recursively?

Reviewer 6: @jmacd yes, it does resolve the question. It seems the behavior of JavaScript SDK is similar to the Go's.

I just like to mention that in the API spec, the callback reentrancy is implementation-defined:

> https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-counter-creation
> [OpenTelemetry API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md#api) authors SHOULD define whether this callback function needs to be reentrant safe / thread safe or not.

And the `state` of the callback is implementation-defined too:

> The API SHOULD provide some way to pass state to the callback. [OpenTelemetry API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md#api) authors MAY decide what is the idiomatic approach (e.g. it could be an additional parameter to the callback function, or captured by the lambda closure, or something else).

So the "outside of a callback" may have different meanings across the implementations. And it would be better to explicitly document this.

Reviewer 7: > @jack-berg is that a reentrant lock? Would it be possible for the callback to invoke itself recursively?

Yes it is reentrant. It would be possible to invoke itself recursively, though probably not advisable.

Author: There's a danger of this PR's scope creeping, but I've moved some of this discussion into a new section of the SDK specification. There were SDK requirements sprinkled into the API requirements at this spot. I've defined "outside the callback" there by way of requiring callbacks to be executed on behalf of a single `MetricReader`.

a callback registered for the instrument.

> Callers SHOULD avoid making duplicate observations from asynchronous

Author: Yes -- the issue of duplicate observations hasn't been identified as clearly as it should, though there is existing language that users SHOULD avoid it. To use your `/proc/stat` example where "cpuN" is the first line of a group of measurements used to identify measurements -- if a buggy kernel produced duplicate lines for "cpu0" let's say, we're not specifying the result.

instrument callbacks, as the result of making duplicate observations

> is not specified.

Reviewer 2: Is something not being specified a reason to include a SHOULD avoid requirement? There are other things that the spec does not specify that don't have a `SHOULD NOT` requirement. Maybe there is another reason why callers SHOULD avoid making duplicate observations.

Author: I refactored this conversation. There's a new sdk.md section on asynchronous instrument requirements.

with the [asynchronous
pattern](https://en.wikipedia.org/wiki/Asynchronous_method_invocation).

Considering duplicate registration of asynchronous instruments, the
implementation MUST support multiple callbacks registered per
identical instrument.

The API MAY provide support for registration of a single callback
having multiple asynchronous instruments, where idiomatic.  The
implementation is REQUIRED to maintain a mapping from every regsitered
callback to the associated instrument(s).

The implementation MUST provide a way to unregister callbacks
associated with asynchronous instruments.

The implementation SHOULD reject observations that are made outside of
a callback registered for the instrument.

> Callers SHOULD avoid making duplicate observations from asynchronous

Reviewer 5: I think this is the right specification level flow here for now.

Alternative:  First observation recorded wins. 
Second alternative: Last observation recorded wins.

Author: This wasn't a change, just a consolidation of existing text below.

There's a lot packed into this tiny discussion, I want to keep it separate. The most easy thing to do for implementations is to apply the natural merge function, which _could_ be defined as the correct outcome if we decided to. Then, the statement would simply be that identical observations will be merged by their natural merge function. If you observe a Sum twice, you'll see the total. If you observe a Gauge twice, you'll see the last value.

I have reasons why this definition might be better, but I don't want to talk about them now.

Author: See #2318. I'd like this to be specified as last-value-wins, ideally, but it's separate from this PR.

with the [asynchronous
pattern](https://en.wikipedia.org/wiki/Asynchronous_method_invocation).

The API MUST support creation of asynchronous instruments by passing
zero or more callback functions to be permanently registered to the
newly created instrument.  Callbacks functions passed to the

> instrument constructor cannot be unregistered.

Reviewer 8: I would form this in a way like:

```
Callbacks functions passed to the instrument constructor MAY not have support to unregistered.
```

Reviewer 8: I would prefer to address this in a separate PR actually.

> newly created instrument.  Callbacks functions passed to the
> instrument constructor cannot be unregistered.

Reviewer 8: ```suggestion
newly created instrument.
```

with the [asynchronous
pattern](https://en.wikipedia.org/wiki/Asynchronous_method_invocation).

The API MUST support creation of asynchronous instruments by passing a
single callback function to be registered to the newly created
instrument.

The API SHOULD support creation of asynchronous instruments
by passing a variable number of callback functions, greater than or
equal to zero, to be registered to the newly created instrument.

The API SHOULD support registration of callbacks bound to

> one or more instruments outside of instrument construcotrs, provided

Reviewer 4: ```suggestion
one or more instruments outside of instrument constructors, provided
```

with the [asynchronous
pattern](https://en.wikipedia.org/wiki/Asynchronous_method_invocation).

The API MUST support creation of asynchronous instruments by passing a

> single callback function to be registered to the newly created

Author: I meant this in the sense "MUST support 1 callback", which is not to preclude creation of instruments with zero callbacks (a SHOULD) or more than one callback (a SHOULD).

instrument.

The API SHOULD support creation of asynchronous instruments
by passing a variable number of callback functions, greater than or
equal to zero, to be registered to the newly created instrument.

> The API MAY support registration of callbacks bound to

Author: Yes, I mean "bound" as in to say that there is always a mapping between instrument and callback. This allows the implementation to enumerate the callbacks that report for any given instrument independently of whether they are registered when the instrument is created. 

The implementation has to have a way to enumerate the callbacks that report for any given instrument because of the requirements about duplicate instrument registration; i.e., we have more than one callback per instrument.

Adding an allowance for multiple instruments changes nothing, i.e., we still have more than one callback per instrument, the only difference is that one callback may appear in the list of callbacks for multiple instruments.

This "binding" so to speak is meant to support exporting instruments with different frequency at some point in the future.

one or more instruments outside of instrument constructors, provided
the API declaratively states which instrument(s) will be be used from
each callback.

> The API SHOULD support unregistration of callbacks, in order to cease

Author: I am now considering a firmer requirement, as it would seem to help:
1. Callback functions passed to instrument constructors cannot be unregistered, because instruments are forever
2. APIs SHOULD support a way to register callbacks after instrument construction as a separate step
3. The API SHOULD support a way to unregister **only those callbacks that were registered after instrument construction**.

> The API SHOULD support unregistration of callbacks, in order to cease

Author: ^^^ This proposal is now applied. To support unregistration of callbacks, you'll have to register them in a separate call than the instrument constructor.

making observations.  Note that unregistration of callbacks does not
undo the effect of instrument registration itself; duplicate
instrument registration conflicts SHOULD still occur for instruments
with no registered callbacks.  The effect of unregistering a callback
more than once is unspecified.

> Callback functions SHOULD NOT take an indefinite amount of time.

Reviewer 9: Since callback functions are registered by the user, should this read more like "Documentation SHOULD be provided indicating to users that callback functions should return quickly and should not make duplicate observations".

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/2350

> Callback functions SHOULD NOT take an indefinite amount of time.

Reviewer: Is a long-running or non-returning callback specifically considered to be undefined behavior?

Author: I'd like implementations to do whatever is realistic here, and don't want to over constrain them. In OTel-Go for example, in a setting where we are committed to timeouts, I would propose to allow each collection interval to use separate goroutines for collection. The export would not block on the callbacks. When the `context.Context` timeout is reached, those goroutines that have not yet finished are on their own, and any measurements they produce will be considered out-of-date on arrival.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/2350

with the [asynchronous
pattern](https://en.wikipedia.org/wiki/Asynchronous_method_invocation).

The API MUST support creation of asynchronous instruments by passing
one or more callback functions to be registered to the newly created
instrument.

The API SHOULD support registration of callback functions to
asynchronous instruments after they are created.

Where the API supports registration of callback functions after
asynchronous instrumentation creation, it MUST return something (e.g.,
a registration handle, receipt or token) to the user that supports

> undoing the effect of callback registation.

Reviewer 8: Would be good as a "recommendation":
1. The "late registered" callbacks have or have not the same signature as the callbacks passed in the constructor? Because you need different signature for the multi-instrument to be backwards compatible.
2. The "late registration" and "unregister" are operations on the "Instrument" or on the "Meter"?

Author: > Because you need different signature for the multi-instrument to be backwards compatible.

I believe we're talking about implementation specifics, but I would say single-instrument callback registration SHOULD use the same signature whether passed to the constructor or registered later.  A multi-instrument callback would be a different signature; this PR is meant to open the door for multi-instrument callbacks by clarifying what is meant by registration and unregistration, but it would be different signature callback functions used for multi-instrument callbacks.

Author: > The "late registration" and "unregister" are operations on the "Instrument"

I was referring to late registration for single-instrument callbacks, so the method to register a single-instrument callback would be an Instrument method.

The unregister method API was not specified, other than that some means be provided. In the examples, which I've updated, the registration token returned by `register_callback()` supports an `unregister()` method directly. I would also be happy with a `Meter` or `MeterProvider` interface. This is not a primary interface, not too important.

  rule](#instrument-unit).
* An optional `description`, following the [instrument description
  rule](#instrument-description).

The API MUST support an idiomatic way to associate one or more

> `callback` functions with each instrument, whether through the

Reviewer 2: ```suggestion
`callback` functions with each Asynchronous Counter instrument, whether through the
```

Author: Done.

#### Asynchronous Counter operations

Asynchronous Counter uses an idiomatic interface for reporting
measurements through a `callback`, which is registered during
[Asynchronous Counter creation](#asynchronous-counter-creation).

The API SHOULD provide support for unregistering callbacks.  For
example:

```python
# Python
class Device:
    """A device with one counter"""

    def __init__(self, meter, x):
        self.x = x
        self.cb = meter.create_observable_counter(name="usage", description="count of items used", self.counter_callback)

    def counter_callback(self, result):
        result.Observe(self.read_counter(), {'x', self.x})

    def read_counter(self):
        return 100  # ...
    
    def stop(self):

>         self.cb.unregister()

Reviewer 8: Can we support "Closing" the instrument/unregister the entire instrument instead? I feel that is much simpler, and is needed anyway.

Author: I think you're suggesting that to unregister or close a specific Instrument "instance" is to unregister the callback(s) that were passed to that instances constructor.

The only difference with what I've tried to specify instead is that it should be clear what we're counting (callback instances) and what lasts forever (instrument instances). It doesn't matter how many times an instrument is registered, it matters how many callbacks were registered as a result. 

The API was marked stable without a requirement that observable instruments return anything. I've made the option to unregister callbacks a true option, so that you aren't required to support unregistration of callbacks passed to instrument constructors, a separate API is permitted. So, without need for an instrument Close() or calling Unregister directly on an asynchronous instrument (whatever that means), you could add a new method, e.g.,

```Python
instrument = meter.create_observable_counter(name="usage", description="count of items used", first_callback)
# there is one callback registered
cb = instrument.register_callback(second_callback)
# there are now two callbacks registered
meter.unregister_callback(cb)
# there is now one callback registered
```
In this example, you can't unregister callbacks passed to the constructor. The API MAY provide another way to register callbacks that can be unregistered.

  rule](#instrument-unit).
* An optional `description`, following the [instrument description
  rule](#instrument-description).

The API MUST support an idiomatic way to associate one or more

> `callback` functions with each instrument, whether through the

Reviewer 2: ```suggestion
`callback` functions with each Asynchronous Gauge, whether through the
```

Author: Fixed.

  rule](#instrument-unit).
* An optional `description`, following the [instrument description
  rule](#instrument-description).

The API MUST support an idiomatic way to associate one or more

> `callback` functions with each instrument, whether through the

Reviewer 2: ```suggestion
`callback` functions with each Asynchronous UpDownCounter, whether through the
```

Author: Well, rephrased actually. I've created a general requirements section for the API because things were being stated 3 times for the async instruments.

* A value
* [`Attributes`](../common/common.md#attributes)

## Multi-instrument asynchronous callback example

Where idiomatic, [the Metrics API supports the use of multiple
instruments from a single registered
callback](#asynchronous-instrument).  Provided there is runtime
support for dynamic scope (e.g., thread-local state, context
variables), this can be achieved by an API similar to the synchronous
instruments.  For example,

```Python
# Python
class Device:
    """A device with one counter"""

    def __init__(self, meter, x):
        self.x = x
        self.usage = meter.create_observable_counter(name="usage", description="count of items used")
        self.pressure = meter.create_observable_gauge(name="pressure", description="force per unit area")

>         self.cb = meter.register_callback([self.usage, self.pressure], self.observe)

Reviewer 6: This example case seems to supersede the batch observers. +1 from me for the idea of "multiple callbacks" instead of a new interface like "batch observer" versus "observable metrics". It is simpler IMO.

Author: Correct. We may take this as a case of a bad prototype leading to a better design. 😁

Reviewer 8: This kind of changes the signature of the callback in a stable API document. Is this not a breaking change? How does languages like Java, C# support this?

Does this have a different signature? Maybe to be consistent with the other callbacks passed in the constructor we pass an array of "results" for all the instruments passed in the register_callback, to avoid adding of a new API called `Observe` on the instrument?

/cc @open-telemetry/java-maintainers @open-telemetry/dotnet-maintainers?

Author: This was given as a MAY in order to prevent breaking stable APIs. You could provide a separate instrument constructor to add this in a stable API.

Author: In any case, I removed multi-instrument callbacks from this PR as discussed in the tracking issue #2280. Another PR can add this easily, it seems.

* A value
* [`Attributes`](../common/common.md#attributes)

## Multi-instrument asynchronous callback example

Where idiomatic, [the Metrics API supports the use of multiple
instruments from a single registered
callback](#asynchronous-instrument).  Provided there is runtime

> support for dynamic scope (e.g., thread-local state, context

Author: Sure, that's good. I'll put your version of this in the future PR. I've removed multi-instrument callbacks from this PR.
